internal/controllers/token_exchange.go

```diff
@@ -210,30 +214,109 @@ func (t *TokenExchangeController) evaluateSacdDoc(c *fiber.Ctx, record *models.P
 		return fiber.NewError(fiber.StatusBadRequest, "Grantee address in permission record doesn't match requester")
 	}
 
-	// Aggregates all the permissions the user has.
-	userPermissions := make(map[string]bool)
+	userPermGrants, cloudEvtGrants := t.userGrantMap(record)
+	for _, agg := range record.Data.Agreements {
+		switch agg.Type {
+		case "cloudevent":
+			if agg.EffectiveAt != nil && !agg.EffectiveAt.IsZero() {
+				if time.Now().Before(*agg.EffectiveAt) {
+					logger.Info().Msgf("agreement not yet in effect: %s", agg.EffectiveAt.String())
+					return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+				}
+			}
+
+			if agg.ExpiresAt != nil && !agg.ExpiresAt.IsZero() {
+				if agg.ExpiresAt.Before(time.Now()) {
+					logger.Info().Msgf("agreement expired: %s", agg.ExpiresAt.String())
+					return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+				}
+			}
+
+			if valid, err := t.validateAssetDID(record.Data.Asset, tokenReq); err != nil || !valid {
+				logger.Err(err).Msgf("failed to validate attestation asset: %s", record.Data.Asset)
+				return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+			}
+
+			if err := t.evaluateCloudEvents(cloudEvtGrants, tokenReq); err != nil {
+				logger.Err(err).Msg("failed to validate request")
+				return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+			}
+
+		case "permissions":
+			// Validate the asset DID if it exists in the record
+			valid, err := t.validateAssetDID(agg.Asset, tokenReq)
+			if err != nil || !valid {
+				logger.Err(err).Msgf("failed to validate asset did %s in permission agreement", agg.Asset)
+				return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+			}
+
+			if err := t.evaluatePermissions(userPermGrants, tokenReq); err != nil {
+				logger.Err(err).Msg("failed to evaluate permissions agreement")
+				return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
+			}
+		}
+	}
+
+	// If we get here, all permission and attestation claims are valid
+	return t.createAndReturnToken(c, tokenReq, grantee)
```

Comment:
We already loop through the agreements in userGrantMap we should not need to do this again.
We can move the validation of assets into userGrantMap since there is already a type switch in there. and then we can have the following for this function
```suggestion
	userPermGrants, cloudEvtGrants := t.userGrantMap(record
                
    if err := t.evaluatePermissions(userPermGrants, tokenReq); err != nil {
		logger.Err(err).Msg("failed to evaluate permissions agreement")
		return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
	}		
	if err := t.evaluateCloudEvents(cloudEvtGrants, tokenReq); err != nil {
		logger.Err(err).Msg("failed to validate request")
		return fiber.NewError(fiber.StatusBadRequest, "failed to validate request")
	}
	return t.createAndReturnToken(c, tokenReq, grantee)
}

	// If we get here, all permission and attestation claims are valid
	return t.createAndReturnToken(c, tokenReq, grantee)
```

Reply:
I kept this in because of the odd situation with validating the asset:
- For the SACD permission grant the asset is inside the permission data but 
- for attestations it's inside the agreements (which is in array).... 

I'm wondering if it makes the most sense the remove the possibility of having a separate asset listed in each attestation agreement (I believe we initially did this for maximum flexibility... but now that I'm working on this change I don't know that there's a concrete use case for it)

Reply:
I think as presently constructed there's no way the assets can differ. I would just not handle this case—the top level should have the asset, that's fine.

Reply:
We seem to be really confusing this amongst ourselves. I think @LorranSutter kinda waffles on this too: in the example [permissions doc](https://github.com/DIMO-network/sacd) there's a top-level asset but none in the agreement, but [the README](https://github.com/DIMO-network/sacd) does not allow for this possibility.

Reply:
The way these are currently put on-chain, there's really no way for the asset to vary, and assets basically have to be attached to an NFT. I would honestly just assume and verify a top-level `asset` directive and forget about the rest for now.

Keep in mind that these two types of agreements might be in the same document.